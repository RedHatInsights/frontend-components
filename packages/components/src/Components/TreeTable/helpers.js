export const sizeCalculator = (rows) => {
    let points = {};

    for (let key = 0; key < rows.length; key++) {
        if (!rows[key].cells) {
            rows[key] = {
                cells: rows[key]
            };
        }

        let currRow = rows[key];
        currRow.level = rows[currRow.treeParent] ? (rows[currRow.treeParent].level + 1) : 0;
        const pointKey = typeof currRow.treeParent === 'undefined' ? 0 : currRow.treeParent + 1;
        if (!points[pointKey]) {
            points[pointKey] = { size: 0 };
        }

        const currPoint = points[pointKey];
        currPoint.size = currPoint.size + 1;
        currRow.point = currPoint;
        currRow.posinset = currPoint.size;
    }

    return rows;
};

export const collapseBuilder = (parentKey = 'treeParent') => (rows, _e, _val, { rowData }) => {
    const currRow = rows[rowData.id];
    const isTreeOpen = !currRow.isTreeOpen;
    const rowsToChange = [
        rows[rowData.id],
        ...isTreeOpen === false ? rows.filter(row => row[parentKey] === rowData.id) : []
    ];
    for (let key = 0; key < rowsToChange.length; key++) {
        rowsToChange[key].isTreeOpen = rowsToChange[key].isTreeOpen === undefined ? undefined : isTreeOpen;
    }

    return rows;
};
