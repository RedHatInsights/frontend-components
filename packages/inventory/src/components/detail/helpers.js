import get from 'lodash/get';

export const redirectToInventoryList = (id, onBackToListClick) => {
    if (onBackToListClick) {
        onBackToListClick();
    } else {
        if (document.referrer) {
            history.back();
        } else {
            location.href = location.pathname.replace(new RegExp(`${[ id ]}.*`, 'g'), '');
        }
    }
};

export const getFact = (path, factDict) => get(factDict, path, undefined);
