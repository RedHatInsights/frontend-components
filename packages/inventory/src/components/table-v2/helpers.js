export const getEnabledFilters = (hideFilters = {}) => ({
    name: !(hideFilters.all && hideFilters.name !== false) && !hideFilters.name,
    stale: !(hideFilters.all && hideFilters.stale !== false) && !hideFilters.stale,
    registeredWith: !(hideFilters.all && hideFilters.registeredWith !== false) && !hideFilters.registeredWith,
    tags: !(hideFilters.all && hideFilters.tags !== false) && !hideFilters.tags
});
