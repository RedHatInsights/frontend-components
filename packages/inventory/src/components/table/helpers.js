import React from 'react';
import { cellWidth, sortable, expandable } from '@patternfly/react-table';
import get from 'lodash/get';
import flatten from 'lodash/flatten';
import TitleColumn from './TitleColumn';
import { Fragment } from 'react';

export const buildCells = (item, columns, extra) => {
    return columns.map(({ key, composed, renderFunc }) => {
        const data = composed ? <Fragment>{TitleColumn(
            composed.map(key => get(item, key, ' ')),
            item.id,
            item,
            extra
        )}</Fragment> : get(item, key, ' ');
        return renderFunc ? <Fragment>{ renderFunc(data, item.id, item, extra) }</Fragment> : data;
    });
};

export const createRows = (rows = [], columns = [], { actions, expandable, noSystemsTable, ...extra } = {}) => {
    if (rows.length === 0) {
        return [{
            cells: [{
                title: noSystemsTable,
                props: {
                    colSpan: columns.length + Boolean(actions)
                }
            }]
        }];
    }

    return flatten(rows.map((oneItem, key) => ([{
        ...oneItem,
        ...oneItem.children && expandable && { isOpen: !!oneItem.isOpen },
        cells: buildCells(oneItem, columns, extra),
        actionProps: {
            'data-ouia-component-id': `${oneItem.id}-actions-kebab`
        }
    }, oneItem.children && expandable && {
        cells: [
            {
                title: typeof oneItem.children === 'function' ? oneItem.children() : oneItem.children
            }
        ],
        parent: key * 2,
        fullWidth: true
    } ]))).filter(Boolean);
};

export const onDeleteFilter = (deleted, currFilter) => {
    const { value: deletedItem } = deleted?.chips?.[0] || {};
    const newFilter = currFilter.filter((item) => item !== deletedItem);
    return newFilter;
};

export const onDeleteTag = (deleted, selectedTags, onApplyTags) => {
    const deletedItem = deleted?.chips?.[0];
    if (selectedTags?.[deleted?.key]?.[deletedItem?.key] !== undefined) {
        selectedTags[deleted?.key][deletedItem?.key] = false;
    }

    onApplyTags && onApplyTags(selectedTags, false);
    return selectedTags;
};

const includesSortable = (transforms) => transforms?.reduce((acc, fn) => acc || fn.toString().includes('onSort:'), false);

export const createColumns = (columns, hasItems, rows, isExpandable) => (
    columns?.map(({ props, transforms, cellFormatters, ...oneCell }) => ({
        ...oneCell,
        transforms: [
            ...transforms || [],
            ...props?.width ? [ cellWidth(props.width) ] : [],
            ...hasItems || rows.length <= 0 || (props && props.isStatic) || transforms?.includes(sortable) || includesSortable(transforms) ? [] : [ sortable ]
        ],
        cellFormatters: [
            ...cellFormatters || [],
            ...isExpandable ? [ expandable ] : []
        ]
    }))
);
