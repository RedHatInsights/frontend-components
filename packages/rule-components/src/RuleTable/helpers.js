import get from 'lodash/get';

export const calculateMeta = (meta) => ({
    itemCount: meta.count,
    perPage: meta.limit || 10,
    page: ((meta.offset || 0) / (meta.limit || 10) + 1),
    ...meta
});

export const calculateActiveFilters = (activeFilter, value, key) => value && value.length > 0 && activeFilter && ({
    category: activeFilter.label,
    type: activeFilter.filterKey,
    chips: activeFilter.filterValues.items && activeFilter.filterValues.items.length > 0 ?
        (Array.isArray(value) ? value : [ value ]).map(oneValue => {
            const currFilter = activeFilter
            .filterValues
            .items
            .find(({ value: currValue }) => `${currValue}` === oneValue);
            return ({
                name: (currFilter || { textual: oneValue }).textual,
                value: oneValue,
                key
            });
        }) : [
            { name: value, key }
        ]
});

export const createRows = (data, columns, expanded, detail) => data.map((oneRule, ruleKey) => ([
    {
        cells: columns.map(({ selector }, cellKey) => ({
            title: typeof selector === 'function' ?
                selector(oneRule, ruleKey, cellKey) :
                get(oneRule, selector, 'Not found')
        })),
        isOpen: !!oneRule.isOpen || expanded.includes(oneRule.rule_id) || expanded.includes(ruleKey),
        ruleId: oneRule.rule_id
    }, detail && {
        parent: ruleKey * 2,
        fullWidth: true,
        noPadding: true,
        cells: [{
            title: typeof detail === 'function' ? detail(oneRule, ruleKey) : detail
        }]
    }
]));
