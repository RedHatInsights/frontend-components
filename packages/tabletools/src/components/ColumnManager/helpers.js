import propTypes from 'prop-types';

export const isSelectedColumn = (selectedColumns, value) => selectedColumns.includes(value);

export const columnShape = propTypes.shape({
  key: propTypes.string.isRequired,
  title: propTypes.string,
  isRequired: propTypes.bool,
});
