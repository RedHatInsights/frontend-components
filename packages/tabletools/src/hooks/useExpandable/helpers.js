import React from 'react';

const childRowForRule = (item, idx, DetailsComponent) => ({
  parent: idx * 2,
  fullWidth: true,
  cells: [{ title: <DetailsComponent item={item} key={'item-' + item.rowId} /> }],
});

// eslint-disable-next-line import/prefer-default-export
export const itemDetailsRow = (item, idx, options) =>
  typeof options?.detailsComponent !== 'undefined' && childRowForRule(item, idx, options.detailsComponent);
