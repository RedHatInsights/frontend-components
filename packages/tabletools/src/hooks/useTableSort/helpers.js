import { sortable } from '@patternfly/react-table';
import uniq from 'lodash/uniq';

const isSortable = (column) => column.sortByProp || column.sortByFunction;

export const addSortableTransform = (columns) =>
  columns.map((column) => ({
    ...column,
    ...(isSortable(column)
      ? {
          transforms: uniq([...(column?.transforms || []), sortable]),
        }
      : {}),
  }));

export const columnOffset = (options = {}) => (typeof options.onSelect === 'function') + (typeof options.detailsComponent !== 'undefined');

const getSortable = (property, item) => {
  if (typeof property === 'function') {
    return property(item);
  } else {
    return item[property];
  }
};

export const orderArrayByProp = (property, objects, direction) =>
  objects.sort((a, b) => {
    const compared = String(getSortable(property, a)).localeCompare(String(getSortable(property, b)));
    if (direction === 'asc') {
      return compared;
    }
    return -compared;
  });

export const orderByArray = (objectArray, orderProp, orderArray, direction) => {
  const sortedObjectArray = orderArray.flatMap((orderKey) => objectArray.filter((item) => item[orderProp] === orderKey));
  return direction !== 'asc' ? sortedObjectArray.reverse() : sortedObjectArray;
};
