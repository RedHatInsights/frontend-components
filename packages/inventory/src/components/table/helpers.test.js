import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import { buildCells, createRows, onDeleteFilter, onDeleteTag, createColumns } from './helpers';

describe('buildCells', () => {
    it('should create cells without renderFunc', () => {
        const data = buildCells({
            first: 'test',
            second: {
                dot: 'dot'
            }
        }, [{ key: 'first' }, { key: 'second.dot' }, { key: 'third' }]);
        expect(data.length).toBe(3);
        expect(data[0]).toBe('test');
        expect(data[1]).toBe('dot');
        expect(data[2]).toBe(' ');
    });

    it('should create cells with renderFunc', () => {
        const renderFunc = jest.fn(() => 'testing');
        const data = buildCells({
            first: 'test',
            second: {
                dot: 'dot'
            }
        }, [{ key: 'first' }, { key: 'second.dot' }, { key: 'third', renderFunc }]);
        expect(data.length).toBe(3);
        expect(data[0]).toBe('test');
        expect(data[1]).toBe('dot');
        expect(data[2]).toMatchObject({ title: 'testing' });
    });
});

describe('createRows', () => {
    const rows = [ ...new Array(5) ].map(() => ({
        first: 'test',
        second: {
            dot: 'dot'
        }
    }));
    const cells = [{ key: 'first' }, { key: 'second.dot' }, { key: 'third' }];
    it('should create empty table', () => {
        const data = createRows();
        const wrapper = mount(data[0].cells[0].title);
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(data[0].cells[0].props.colSpan).toBe(0);
    });

    it('should create empty table with actions', () => {
        const data = createRows([], cells, { actions: [] });
        const wrapper = mount(data[0].cells[0].title);
        expect(toJson(wrapper)).toMatchSnapshot();
        expect(data[0].cells[0].props.colSpan).toBe(4);
    });

    it('should create regular table', () => {
        const data = createRows(rows, cells);
        expect(data.length).toBe(rows.length);
        expect(data[0].cells[0]).toBe('test');
        expect(data[0].cells[1]).toBe('dot');
        expect(data[0].cells[2]).toBe(' ');
    });

    describe('expandable', () => {
        it('should create collapsed table', () => {
            const data = createRows(
                rows.map(item => ({ ...item, children: <div>Something</div> })),
                cells,
                { expandable: true }
            );
            expect(data[1].parent).toBe(0);
            expect(data[3].parent).toBe(2);
            expect(data.length).toBe(rows.length * 2);
        });

        it('should create expanded table with function children', () => {
            const data = createRows(
                rows.map(item => ({ ...item, children: () => 'something', isOpen: true })),
                cells,
                { expandable: true }
            );
            expect(data[0].isOpen).toBe(true);
            expect(data.length).toBe(rows.length * 2);
        });
    });
});

describe('onDeleteFilter', () => {
    const filter = [ 'something' ];
    it('should delete filter', () => {
        const data = onDeleteFilter({
            chips: [{
                value: 'something'
            }]
        }, filter);
        expect(data.length).toBe(0);
    });

    it('should not delete filter', () => {
        const data = onDeleteFilter({
            chips: [{
                value: 'wrong'
            }]
        }, filter);
        expect(data.length).toBe(1);
        expect(data).toMatchObject(filter);
    });

    it('should not delete filter if no delete filter', () => {
        const data = onDeleteFilter(undefined, filter);
        expect(data.length).toBe(1);
        expect(data).toMatchObject(filter);
    });
});

describe('onDeleteTag', () => {
    const selectedTags = {
        some: {
            tag: true
        }
    };
    it('should call onDeleteTag with updated value', () => {
        const onApplyTags = jest.fn();
        const data = onDeleteTag({
            chips: [{
                key: 'tag'
            }],
            key: 'some'
        }, selectedTags, onApplyTags);
        expect(onApplyTags).toHaveBeenCalled();
        expect(data.some.tag).toBe(false);
    });

    it('should call onDeleteTag without updated value', () => {
        const data = onDeleteTag({}, selectedTags);
        expect(data).toMatchObject(selectedTags);
    });
});

describe('createColumns', () => {
    it('should create regular columns with custom transforms and cellFormatters', () => {
        const data = createColumns([{
            transforms: [ () => undefined ],
            cellFormatters: [ () => undefined ]
        }, { data: 'something' }], false, [ 'something' ]);
        expect(data[0].transforms.length).toBe(2);
        expect(data[0].cellFormatters.length).toBe(1);
        expect(data[1].transforms.length).toBe(1);
        expect(data[1]).toMatchObject({ data: 'something' });
    });

    it('should create expandable columns', () => {
        const data = createColumns([{ data: 'something' }], false, [ 'something' ], true);
        expect(data[0].cellFormatters.length).toBe(1);
    });

    it('should create columns with width', () => {
        const data = createColumns([{ data: 'something', props: { width: 10 } }], false, [ 'something' ]);
        expect(data[0].transforms.length).toBe(2);
    });

    describe('non sortable', () => {
        it('should respect isStatic prop', () => {
            const data = createColumns([{ data: 'something', props: { isStatic: true } }], false, [ 'something' ]);
            expect(data[0].transforms.length).toBe(0);
        });

        it('should respect items with 0 length', () => {
            const data = createColumns([{ data: 'something' }], true, [ 'something' ]);
            expect(data[0].transforms.length).toBe(0);
        });
    });
});
