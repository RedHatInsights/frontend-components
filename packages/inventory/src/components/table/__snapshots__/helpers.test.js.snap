// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCells should create cells with renderFunc 1`] = `"testing"`;
